### ISSUE TITLE
Bug when using a CustomCentroidBondForce with platform CPU, but not with CUDA or OpenCL platforms

### ISSUE THREAD

**Issue by FHedin** [pandegroup/openmm#1848]
Dear OMM developers,

I am using OpenMM version 7.1.1 (compiled myself, using the source corresponding to the git tag 7.1.1).

I want to perform Langevin dynamics of the protein FKPB bound to the DMSO ligand (PDB id : 1d7h), using Amber files  and the OBC2 implicit solvent model.

I have added a CustomCentroidBondForce, using as the two groups : 

- The centre of mass (COM) of the DMSO ligand
- The COM of the residues' heavy atoms forming the binding pocket

I use for this the following code (full script + other files are provided in a zip archive below)

```
...

print("Creating CustomCentroidBondForce...")

# we put a constraint to keep the DMSO ligand at less than 20 AA (2.0 nm) from the CM of the residues defining the binding pocket
force = CustomCentroidBondForce(2,'50*max(0,distance(g1,g2)-2.0)^2')
# DMSO atomID range (when starting from 0) is atoms 1663 to 1672
dmso = range(1663,1672,1)
# pocket indices when starting from 0 are:
pocket = [394,396,398,400,404,406,407,561,563,565,568,570,571,713,715,717,720,
	  723,726,729,733,734,868,870,872,874,878,881,885,886,887,889,891,894,
	  897,900,902,903,906,909,910,942,944,946,949,952,953,954,955,956,1287,
	  1289,1292,1293,1525,1527,1529,1532,1533,1534,1535,1536]
# make groups from atom lists
g_dmso   = force.addGroup(dmso)
g_pocket = force.addGroup(pocket)
# register a virtual CentroidBond between the two above defined groups
force.addBond([g_dmso,g_pocket])

system.addForce(force)

...

```

When using the CUDA or OpenCL platforms, it does what I expect, i.e. keeping the DMSO at no more than 2 nm from the COM of the pocket when the ligand switches to the unbound state (after a few ns of simulation). The output of the dynamics is as following : 
```

Creating System...
Creating CustomCentroidBondForce...
Creating Langevin integrator...
Creating Platform and Simulation ...
Minimizing...
Dynamics...
#"Progress (%)"	"Step"	"Time (ps)"	"Potential Energy (kJ/mole)"	"Kinetic Energy (kJ/mole)"	"Total Energy (kJ/mole)"	"Temperature (K)"	"Speed (ns/day)"	"Time Remaining"
10.0%	500	1.0	-11686.0598519	4857.42816602	-6828.63168583	279.594942644	0	--
20.0%	1000	2.0	-11426.8335154	5298.29824603	-6128.53526938	304.971549466	90.2	0:07
30.0%	1500	3.0	-11288.4141498	5364.23819061	-5924.17595923	308.767071374	89.8	0:06
40.0%	2000	4.0	-11341.8010511	5521.65452157	-5820.14652949	317.828000022	89.7	0:05
50.0%	2500	5.0	-11446.400135	5319.64976627	-6126.75036875	306.200549026	90.2	0:04
60.0%	3000	6.0	-11344.271714	5317.49947265	-6026.77224133	306.076777515	90.6	0:03
70.0%	3500	7.0	-11562.6428486	5371.44162457	-6191.20122402	309.181703076	90.5	0:02
80.0%	4000	8.0	-11354.2005238	5257.29527924	-6096.9052446	302.61140707	90.3	0:01
90.0%	4500	9.0	-11413.8497478	5218.07355233	-6195.7761955	300.353793348	90.3	0:00
100.0%	5000	10.0	-11482.8734358	5316.2854329	-6166.5880029	306.006897043	90.4	0:00
```

However when running using the CPU platform I get the following (x5 larger potential energy) : 

```
Creating System...
Creating CustomCentroidBondForce...
Creating Langevin integrator...
Creating Platform and Simulation ...
Minimizing...
Dynamics...
#"Progress (%)"	"Step"	"Time (ps)"	"Potential Energy (kJ/mole)"	"Kinetic Energy (kJ/mole)"	"Total Energy (kJ/mole)"	"Temperature (K)"	"Speed (ns/day)"	"Time Remaining"
10.0%	500	1.0	-54996.4808482	5182.30898902	-49814.1718592	298.295174942	0	--
20.0%	1000	2.0	-54844.4616318	5250.24593825	-49594.2156935	302.205645004	2.02	5:42
30.0%	1500	3.0	-55092.4538069	5386.28528774	-49706.1685191	310.036108537	2.09	4:49
40.0%	2000	4.0	-55082.677836	5190.48822222	-49892.1896138	298.765973924	2.15	4:00
50.0%	2500	5.0	-55277.242945	5247.32543443	-50029.9175106	302.037539976	2.21	3:15
60.0%	3000	6.0	-55136.6001528	5338.4815575	-49798.1185953	307.284512269	2.26	2:33
70.0%	3500	7.0	-55121.9038036	5387.77225554	-49734.131548	310.121698825	2.3	1:52
80.0%	4000	8.0	-55259.6690671	5236.86695706	-50022.8021101	301.435547053	2.33	1:14
90.0%	4500	9.0	-55423.1948491	5479.09105924	-49944.1037898	315.378035061	2.37	0:36
100.0%	5000	10.0	-55189.118428	5464.63577993	-49724.4826481	314.545985085	2.4	0:00
```
The following is a VMD snapshot showing how distorted is the protein before the dynamics, just after minimisation : white corresponds to the input coordinates, green to a minimised CUDA system, red to a CPU minimised system : 

![vis](https://user-images.githubusercontent.com/3983545/27594463-19416c2c-5b5a-11e7-9c33-6ac96e396741.png)

The following zip archives contains inputs, logs, pdbs and trajectories.

[bug_omm_customCentroid.zip](https://github.com/pandegroup/openmm/files/1105727/bug_omm_customCentroid.zip)

I hope this post will be helpful !

Regards,

Florent Hédin

**Comment by peastman** [comment]
I'm still tracking down what's going on, but it's nothing to do with the CustomCentroidBondForce.  This happens even if I remove it.  The actual difference in energy is coming from the CustomGBForce.

**Comment by FHedin** [comment]
Oh OK, I never had such a problem before and I first noticed it after enabling a CustomCentroidBondForce, so I suspected the wrong force.

I tried to simulate the same system but this time using charmm36 + cgenff parameters : I got the same problem for the CPU platform.

I also tried to remove the saltConcentration parameter that I have never used before, but still the same issue.

**Comment by FHedin** [comment]
So some more remarks concerning the option implicitSolventSaltConc for createSystem(...) : 

- When building the system from AMBER files, omitting implicitSolventSaltConc (so using the default of 0) appears to solve the problem and give a proper potential energy : 

```
#"Progress (%)"	"Step"	"Time (ps)"	"Potential Energy (kJ/mole)"	"Kinetic Energy (kJ/mole)"	"Total Energy (kJ/mole)"	"Temperature (K)"	"Speed (ns/day)"	"Time Remaining"
10.0%	500	1.0	-11511.8571467	5126.5700135	-6385.28713316	295.086823705	0	--
20.0%	1000	2.0	-11283.594536	5343.36652779	-5940.22800824	307.565692544	12.6	0:54
30.0%	1500	3.0	-11472.5409942	5566.94006202	-5905.60093217	320.434648572	12.5	0:48
40.0%	2000	4.0	-11238.9348976	5136.81131779	-6102.12357983	295.676316084	12.5	0:41
50.0%	2500	5.0	-11065.4355109	5505.18848297	-5560.24702797	316.880210171	12.6	0:34
60.0%	3000	6.0	-11061.3890597	5491.65126006	-5569.73779962	316.101003782	12.1	0:28
70.0%	3500	7.0	-11214.0907132	5369.13039742	-5844.96031575	309.048668186	9.69	0:26
80.0%	4000	8.0	-11468.6524016	5263.94058931	-6204.71181224	302.993913002	7.59	0:22
90.0%	4500	9.0	-11129.6250432	5493.07210824	-5636.55293499	316.182788206	6.63	0:13
100.0%	5000	10.0	-11018.1332209	5209.50898486	-5808.62423606	299.860814415	6.03	0:00

```
- When building the system from CHARMM files, omitting implicitSolventSaltConc (so using the default of 0) does not change much : 

```
#"Progress (%)"	"Step"	"Time (ps)"	"Potential Energy (kJ/mole)"	"Kinetic Energy (kJ/mole)"	"Total Energy (kJ/mole)"	"Temperature (K)"	"Speed (ns/day)"	"Time Remaining"
10.0%	500	1.0	-49635.2408703	4915.80434091	-44719.4365294	282.955091825	0	--
20.0%	1000	2.0	-49690.0924808	5293.4398893	-44396.6525915	304.691901075	2.05	5:37
30.0%	1500	3.0	-49601.2747881	5403.09141793	-44198.1833702	311.003474156	2.14	4:42
40.0%	2000	4.0	-49745.8713403	5539.37423061	-44206.4971097	318.8479515	2.23	3:52
50.0%	2500	5.0	-50103.4399227	5584.7139065	-44518.7260162	321.457716101	2.3	3:07
60.0%	3000	6.0	-50184.6643158	5445.25860906	-44739.4057068	313.430629635	2.36	2:26
70.0%	3500	7.0	-49908.1613348	5463.25326893	-44444.9080658	314.46640736	2.42	1:47
80.0%	4000	8.0	-50258.6600975	5570.17047901	-44688.4896185	320.62059229	2.47	1:10
90.0%	4500	9.0	-50188.4693654	5475.66679145	-44712.8025739	315.180933236	2.52	0:34
100.0%	5000	10.0	-50193.9553215	5452.86578885	-44741.0895326	313.868501061	2.57	0:00
```

**Comment by peastman** [comment]
I just sent a PR with the fix: #1850.
